tidy3d/plugins/microwave/array_factor.py

```diff
@@ -761,28 +875,394 @@ def array_factor(
+        Sample weights from the circular Taylor taper at specified radial positions.
+
+        Parameters
+        ----------
+        p_vec : ArrayLike
+            1D array of radial sampling points in the range [0, π].
+
+        Returns
+        -------
+        g_p_norm : ArrayLike
+            1D array of Taylor taper weights evaluated at the points in ``p_vec``.
+        """
+
+        # get locations J1(np.pi * mu)=0
+        mus = jn_zeros(1, self.nbar) / np.pi
+
+        B_m = self._get_exp_weights(mus=mus)
+
+        J = j0(np.outer(mus[0:-1], p_vec))
+
+        g_p_c = 1 + B_m @ j0(0 * mus[0:-1])
+        g_p = 1 + B_m @ J
+
+        g_p_norm = g_p / g_p_c
+
+        return g_p_norm
+
+
+class AbstractTaper(Tidy3dBaseModel, ABC):
+    """Abstract taper class provides an interface for taper of Array antennas."""
+
+    window: AbstractWindow = pd.Field(
+        ..., title="Window object", description="Window type used to taper array antenna."
+    )
+
+    @abstractmethod
+    def amp_multipliers(
+        self, array_size: tuple[PositiveInt, PositiveInt, PositiveInt]
+    ) -> tuple[np.ndarray, ...]:
+        """
+        method ``amp_multiplier()`` computes taper amplitudes for phased array antennas.
+
+        Parameters:
+        ----------
+        array_size: tuple[PositiveInt, PositiveInt, PositiveInt]
+            A tuple of array size along x,y, and z axes.
+        """
+
+
+class RectangularTaper(AbstractTaper):
+    """Class for rectangular taper."""
+
+    @pd.validator("window", check_fields=False)
+    def validate_window(cls, v):
+        if isinstance(v, TaylorTaperc):
+            raise ValueError(
+                "`TaylorTaperc` is not supported by `RectangularTaper`. "
+                "Use `RadialTaper` if you want to use a Taylor taper."
+            )
+        return v
+
+    def amp_multipliers(
+        self, array_size: tuple[PositiveInt, PositiveInt, PositiveInt]
+    ) -> tuple[ArrayLike, ArrayLike, ArrayLike]:
+        """
+        Method ``amp_multipliers()`` computes rectangular taper amplitude for phased array antennas.
+
+        Parameters:
+        ----------
+        array_size: tuple[PositiveInt, PositiveInt, PositiveInt]
+            A tuple of array size along x,y, and z axes.
+
+        Returns:
+        --------
+        tuple[ArrayLike, ArrayLike, ArrayLike]
+            a tuple of three 1D numpy arrays with taper amplitudes along x,y, and z axes.
+        """
+
+        amps = tuple(self.window.get_weights(dim) if dim is not None else 1.0 for dim in array_size)
+
+        return amps
+
+
+class RadialTaper(AbstractTaper):
+    """Class for radial taper."""
+
+    @pd.validator("window", check_fields=False)
+    def validate_window(cls, v):
+        if not isinstance(v, TaylorTaperc):
+            raise ValueError(
+                f"`{type(v).__name__}` is not supported by `RadialTaper`. "
+                f"Use `RectangularTaper` for this type of window."
+            )
+        return v
```

Comment:
instead of creating a custom validator it's better to simply override window with specifying which types are allowed. That is:
```
    window: TaylorTaperc = pd.Field(
        ..., title="Window object", description="Window type used to taper array antenna."
    )
```
this will automatically check that window is only `TaylorTaperc`, and another advantage is that in API docs it will also show that only this window is allowed